### DISCUSSION TITLE
Ubuntu

### DISCUSSION THREAD

**Discussion by casper-none** [3urobeat/steam-comment-service-bot#216] [Q&A] [answered]
Can I run this script on Ubuntu?
I made systemd so that the script runs in the background but I get this error
![image](https://github.com/3urobeat/steam-comment-service-bot/assets/63936597/e1910572-65bd-4e4f-bb97-17612db3f2f4)
This is what the file looks like
```
[Unit]
Description=Comment Bot
After=network.target

[Service]
ExecStart=/usr/bin/node /home/comms/commentbot/start.js
Restart=on-success
RestartSec=5s

[Install]
WantedBy=multi-user.target

```

**Comment by 3urobeat** [comment]
Hi, yes of course! I myself am running Linux exclusively for years now.  
Not sure why it is detecting a corrupted file, as long as you didn't make any modifications it should work just fine.  
I would need to test later if it is maybe caused by starting it using an absolute path instead of directly from the bot's folder; I haven't tested this in quite some time.  
I'll reply back when I know more, I'm currently at work

**Reply by casper-none** [reply]
ok, im waiting. Thx

**Reply by 3urobeat** [reply]
Alright, this is indeed caused by using an absolute path; sorry for missing something like this lol
I have opened a new issue to keep track of this bug: #217
In the meantime please make sure to start the bot from the bot's folder directly - it might be possible to chain the systemd exec command, let me check real quick

**Reply by 3urobeat** [reply, marked as answer]
Alright, found a better solution for now - simply define a WorkingDirectory.  
Your service file should look like this:
```
[Unit]
Description=Comment Bot
After=network.target

[Service]
WorkingDirectory=/home/comms/commentbot
ExecStart=/usr/bin/node start.js
Restart=on-success
RestartSec=5s

[Install]
WantedBy=multi-user.target
```

I'll update #217 once the bug has been fixed :)

**Reply by casper-none** [reply]
I'm using systemctl `systemctl start comms@comms` to run a script in the background via systemd, but I'm getting this
```
● comms@comms.service - Comment Bot
     Loaded: loaded (/etc/systemd/system/comms@.service; disabled; vendor preset: enabled)
     Active: active (running) since Wed 2023-11-08 01:37:33 MSK; 11s ago
   Main PID: 58857 (CommentBot)
      Tasks: 18 (limit: 1013)
     Memory: 117.8M
        CPU: 4.421s
     CGroup: /system.slice/system-comms.slice/comms@comms.service
             ├─58857 CommentBot "" "" "" "" "" "" "" "" "" "" "" ""
             └─58864 CommentBot "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" >
Nov 08 01:37:44 2000561.twcnet node[58864]: [193B blob data]
Nov 08 01:37:44 2000561.twcnet node[58864]: [4B blob data]
Nov 08 01:37:44 2000561.twcnet node[58864]: [140B blob data]
Nov 08 01:37:44 2000561.twcnet node[58864]: [4B blob data]
Nov 08 01:37:44 2000561.twcnet node[58864]: [176B blob data]
Nov 08 01:37:44 2000561.twcnet node[58864]: [4B blob data]
Nov 08 01:37:44 2000561.twcnet node[58864]: [195B blob data]
Nov 08 01:37:44 2000561.twcnet node[58864]: [4B blob data]
Nov 08 01:37:45 2000561.twcnet node[58864]: [187B blob data]
Nov 08 01:37:45 2000561.twcnett node[58864]: [4B blob data]
```
When I just run the script in the terminal it works but when I close the terminal the script exits

**Reply by 3urobeat** [reply]
Well, I can't exactly tell you why your systemd script is failing here, my experience with writing systemd service files is limited anyway :'D  
Starting the bot using the service file I posted above which specifies a WorkingDirectory works flawlessly for me, I can close any and all terminals just fine  

Please try the workaround I posted and select it as an answer should that work for you as well

**Reply by casper-none** [reply]
Can I ask one more thing? I get an error saying that 2 accounts have a limit and I need to add them as friends. When I add them it says the same thing

**Reply by 3urobeat** [reply]
Sure, always feel free to ask :)  
Accounts which have spent less than $5 on the Steam Store are "limited", meaning they can't do a few things on Steam, one thing being commenting on non-friend profiles. When adding an account, Steam has multiple friendship statuses and it might not yet have switched to the correct one - I have noticed this recently too. Either that or there is a new bug since v2.14 (either in my bot or in one of the libraries) as it worked flawlessly previously.  
Have you waited a moment and restarted the bot since adding those accounts? 

Edit: Restarting the bot shouldn't be necessary, this is normally instant

**Reply by casper-none** [reply]
Now I'll try again

**Reply by casper-none** [reply]
No bro, I run your file but I get the same error
```
Nov 08 20:36:45 2000925-fa44561.twc1.net systemd[1]: Started Comment Bot.
Nov 08 20:36:45 2000925-fa44561.twc1.net root[70511]: [11B blob data]
Nov 08 20:36:45 2000925-fa44561.twc1.net root[70511]: [52B blob data]
Nov 08 20:36:45 2000925-fa44561.twc1.net root[70518]: [9B blob data]
Nov 08 20:36:46 2000925-fa44561.twc1.net root[70518]: [1.1K blob data]
```

**Reply by 3urobeat** [reply]
Systemd still shutting down your service? Problem is probably on your end, it works fine for me, sorry. The "blob data" log is a known issue, just check the `output.txt` file in the bot's folder, it contains the full log.  

If friendship status is still a problem - I'm afraid I can't do anything about this right now; either Steam reports a wrong friendship status or something else is suddenly broken, I however didn't change anything related to that feature iirc

**Reply by casper-none** [reply]
The file `output.txt` contains the last log when I ran the script myself without using systemd

**Reply by 3urobeat** [reply]
Using the systemd service file I provided in my comment which includes the `WorkingDirectory`??  
That should be impossible as systemd now executes the bot just like you would manually.  
You maybe need to reload your service file after making changes: `sudo systemctl daemon-reload`

Edit: Should that command fail on Ubuntu, try `sudo systemctl --user daemon-reload`. Also don't forget to restart the service afterwards.

**Reply by casper-none** [reply]
When I use
```
[Unit]
Description=Comment Bot
After=network.target

[Service]
WorkingDirectory=/root/comms
ExecStart=/usr/bin/node start.js
Restart=on-success
RestartSec=5s

[Install]
WantedBy=multi-user.target
```
Then I get this error:
```
Nov 08 20:59:12 2000925-fa44561.twc1.net systemd[1]: Started Comment Bot.
Nov 08 20:59:13 2000925-fa44561.twc1.net node[73079]: [11B blob data]
Nov 08 20:59:13 2000925-fa44561.twc1.net node[73079]: [52B blob data]
Nov 08 20:59:13 2000925-fa44561.twc1.net node[73089]: [9B blob data]
Nov 08 20:59:14 2000925-fa44561.twc1.net node[73089]: [1.2K blob data]
```
But when I use a slightly different option:
```
[Unit]
Description=Comment Bot
Wants=network.target
After=network.target

[Service]
User=root
Group=root
ExecStart=/usr/bin/node /root/comms/start.js
Restart=on-success
RestartSec=5s
SyslogIdentifier=root

[Install]
WantedBy=multi-user.target
```
I get this:
```
● comms.service - Comment Bot
     Loaded: loaded (/etc/systemd/system/comms.service; linked; vendor preset: enabled)
     Active: active (running) since Wed 2023-11-08 21:02:30 MSK; 1s ago
   Main PID: 73275 (CommentBot)
      Tasks: 25 (limit: 1013)
     Memory: 28.0M
        CPU: 1.061s
     CGroup: /system.slice/comms.service
             ├─73275 CommentBot "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" "" ""
             └─73285 /usr/bin/node --max-old-space-size=2048 --optimize-for-size ./src/controller/controller.js /root/comms/src 1699466550329

Nov 08 21:02:30 2000925-fa44561.twc1.net systemd[1]: Started Comment Bot.
Nov 08 21:02:30 2000925-fa44561.twc1.net root[73275]: [11B blob data]
Nov 08 21:02:30 2000925-fa44561.twc1.net root[73275]: [52B blob data]
Nov 08 21:02:30 2000925-fa44561.twc1.net root[73285]: warn | It looks like file ./src/controller/helpers/logger.js is corrupted. Trying to pull new file from GitHub...
Nov 08 21:02:30 2000925-fa44561.twc1.net root[73285]: info | Pulling: https://raw.githubusercontent.com/3urobeat/steam-comment-service-bot/master/src/controller/helpers/logger.js
```

**Reply by 3urobeat** [reply]
> Then I get this error:

The "blob data" output there isn't an "error", systemd just somehow fails to display the actual text.  
Whatever, just take a look at the output.txt file in the bot's folder, it should contain the full log, without any problems.  
It works for me without issues.

> But when I use a slightly different option:

You are missing the `WorkingDirectory` again. This causes the data integrity check to fail because the execution path is now wrong. Exactly the same issue as in your very first example
I'm not sure why the log is now showing but whatever, doesn't really matter as you can just check the output.txt file.  
I also wouldn't advise to run the bot as `root` on your host machine!

**Reply by 3urobeat** [reply]
Please just use the file I have posted in my third comment, reload systemd using the daemon-reload command I just posted and it should work fine and your output should appear in your output.txt file. Make sure to reopen that file if your text editor does not support automatic refreshes

**Reply by 3urobeat** [reply]
Otherwise - if you for whatever reason can't get your systemd service to work, start the bot manually and put it into the background:
```
cd path/to/your/bot/folder
node start.js
PRESS CTRL+Z
bg
disown
```
This will also work, you just need to start the bot manually should you restart your machine. It will auto-restart should it crash because my bot can handle that itself.

**Reply by casper-none** [reply]
How can I turn this off?

**Reply by 3urobeat** [reply]
Turn what off?  
Auto restarting? You cannot - the bot will restart if the main process crashes. If you stop the bot like normal it won't restart.  
The output file? You cannot and why would you, it is there so that you can keep track of what's happening, even if you closed your stdout (e.g. terminal).  
You can customize nearly everything, those are two of the very few things that are just meant to work like this. I cannot add a setting for turning off the log file as the config files are loaded after the first few log calls have already been made.

I think this discussion is spiraling out of control a bit, have you managed to get the bot running now using either a systemd file or by starting the bot manually and putting it into the background? Both options work for me flawlessly, I have tested them multiple times now.

**Reply by casper-none** [reply]
systemd still doesn't work, I don't understand what's wrong. Starting the bot in manual mode and switching it to background works

**Reply by casper-none** [reply]
Bro, it works.
I thought it was a bug so when I started it I immediately turned it off
```
Nov 09 15:56:32 2000925-fa44561.twc1.net node[5061]: [164B blob data]
Nov 09 15:56:32 2000925-fa44561.twc1.net node[5061]: [4B blob data]
Nov 09 15:56:32 2000925-fa44561.twc1.net node[5061]: [164B blob data]
Nov 09 15:56:32 2000925-fa44561.twc1.net node[5061]: [4B blob data]
```
But now I started everything the same way and saw that the bot had appeared on the network. It turns out it worked but I thought there was an error. 
Thank you for your help and your time.

**Reply by 3urobeat** [reply]
That's great to hear! I'm glad it is working for you now.  
Please select my third comment containing the fixed systemd service file as an answer so that this thread is marked as resolved :)  
Feel free to open another thread if you run into any other issues ^^
